doc/designs/quic-design/server/quic-server-api.md

```diff
@@ -0,0 +1,521 @@
+QLSO (QUIC_LISTENER/QUIC_PORT)              Event Leader, Port Leader
+    [QUIC_ENGINE]
+    QCSO (QUIC_CONNECTION/QUIC_CHANNEL)
+        QSSO (QUIC_XSO/QUIC_STREAM)
+```
+
+Here we see that the QLSO now has the Event Leader role, as it is the most
+senior object visible publicly as a constructed API object. Note that a
+QUIC_ENGINE will be created internally to allow the QLSO to service the Event
+Leader role, but the corresponding QUIC_DOMAIN APL object will not exist, as
+this is not API-visible.
+
+Example 3 - Explicitly Created Single Client QCSO:
+
+```text
+QCSO (QUIC_CONNECTION/QUIC_CHANNEL)         Event Leader, Port Leader
+    [QUIC_ENGINE]
+    [QUIC_PORT]
+    QSSO (QUIC_XSO/QUIC_STREAM)
+```
+
+Here we see that the QCSO now has the Port Leader role. Note that a QUIC_PORT
+will be created internally to allow the QCSO to service the Port Leader role,
+but the corresponding QUIC_LISTENER APL object will not exist, as this is not
+API-visible. Similarly a QUIC_ENGINE is spun up to handle the Event Leader role,
+as above.
+
+It should be emphasised that everything in this section is largely academic and
+intended to demonstrate potential future directions and the adaptability of the
+design to future evolution. There is no immediate plan to implement a QDSO
+object type in the APL. However it will be viable to do so if we want to support
+multiple sockets per domain someday.
+
+| Abbr. | libssl APL Object | QUIC Core Object | Seniority   | Potential Roles |
+|-------|-------------------|------------------|-------------|-----------------|
+| QDSO  | `QUIC_DOMAIN`     | `QUIC_ENGINE`    | 1 (Highest) | EL              |
+| QLSO  | `QUIC_LISTENER`   | `QUIC_PORT`      | 2           | EL PL           |
+| QCSO  | `QUIC_CONNECTION` | `QUIC_CHANNEL`   | 3           | EL PL           |
+| QSSO  | `QUIC_XSO`        | `QUIC_STREAM`    | 4           |                 |
+
+### Polling
+
+We will support polling all QUIC object types (QLSOs, QCSOs, QSSOs) using a new
+polling API. This is the subject of its own design and will be discussed in more
+detail in a separate design document.
+
+### Impact of `SSL_METHOD` selection
+
+The interaction of our various QCSO and QLSO construction APIs in the APL,
+and the choice of the `SSL_METHOD` for a `SSL_CTX`, must also be considered.
+The following interactions are chosen:
+
+```text
+SSL_new(QUIC_client_method)             → QCSO(Client)
+SSL_new(QUIC_client_thread_method)      → QCSO(Client)
+SSL_new(QUIC_server_method)             → (not supported; error)
+
+SSL_new_listener(QUIC_client_method)        → QLSO(Client)   (implicit NO_ACCEPT)
+SSL_new_listener(QUIC_client_thread_method) → QLSO(Client)   (implicit NO_ACCEPT)
+SSL_new_listener(QUIC_server_method)        → QLSO(Server + Client)
+
+SSL_new_from_listener(QUIC_client_method)           → QCSO (Client)
+SSL_new_from_listener(QUIC_client_thread_method)    → QCSO (Client)
+SSL_new_from_listener(QUIC_server_method)           → QCSO (Client)
```

Comment:
I find this information quite confusing, since you seem to show SSL_new() accepting a method (when it takes an SSL_CTX), and SSL_new_listener() also accepting a method (but the text earlier said it took an SSL).

I think I understand what you are trying to show, but it is not clear.

Reply:
Ah sorry. Of course I mean an SSL_CTX with the given method set, etc.